core/src/main/java/org/infinispan/notifications/cachelistener/CacheNotifierImpl.java

```diff
@@ -1070,51 +1085,89 @@ public void addListener(Object listener, KeyFilter<? super K> filter, ClassLoade
       }
 
       // If we have a segment listener handler, it means we have to do initial state
-      QueueingSegmentListener handler = segmentHandler.remove(generatedId);
+      QueueingSegmentListener<K, V, ? extends Event<K, V>> handler = segmentHandler.remove(generatedId);
       if (handler != null) {
          if (trace) {
             log.tracef("Listener %s requests initial state for cache", generatedId);
          }
-         CacheStream entryStream = cache.wired().withEncoding(keyConversion.getEncoderClass(), valueConversion.getEncoderClass()).cacheEntrySet().stream();
-         Stream<CacheEntry<K, V>> usedStream = entryStream.segmentCompletionListener(handler);
+         Queue<IntermediateOperation> intermediateOperations = new ArrayDeque<>();
+
+         if (keyDataConversion != DataConversion.IDENTITY_KEY && valueDataConversion != DataConversion.IDENTITY_VALUE) {
+            intermediateOperations.add(new MapOperation(EncoderEntryMapper.newCacheEntryMapper(
+                  keyDataConversion, valueDataConversion, entryFactory)));
+         }
 
          if (filter instanceof CacheEventFilterConverter && (filter == converter || converter == null)) {
             // Hacky cast to prevent other casts
-            usedStream = CacheFilters.filterAndConvert(usedStream,
-                  new CacheEventFilterConverterAsKeyValueFilterConverter<>((CacheEventFilterConverter<K, V, V>) filter));
+            intermediateOperations.add(new FlatMapOperation(CacheFilters.flatMap(
+                  new CacheEventFilterConverterAsKeyValueFilterConverter<>((CacheEventFilterConverter) filter))));
```

Comment:
This also works and avoids all the unchecked warnings:

```java
         Queue<IntermediateOperation<?, ?, ?, ?>> intermediateOperations = new ArrayDeque<>();

         if (keyDataConversion != DataConversion.IDENTITY_KEY && valueDataConversion != DataConversion.IDENTITY_VALUE) {
            intermediateOperations.add(new MapOperation<>(EncoderEntryMapper.newCacheEntryMapper(
                  keyDataConversion, valueDataConversion, entryFactory)));
         }

         if (filter instanceof CacheEventFilterConverter && (filter == converter || converter == null)) {
            intermediateOperations.add(new FlatMapOperation<>(CacheFilters.flatMap(
                  new CacheEventFilterConverterAsKeyValueFilterConverter<>((CacheEventFilterConverter<?, ?, ?>) filter))));
         } else {
            if (filter != null) {
               intermediateOperations.add(new FilterOperation<>(CacheFilters.predicate(
                     new CacheEventFilterAsKeyValueFilter<>(filter))));
            }
            if (converter != null) {
               intermediateOperations.add(new MapOperation<>(CacheFilters.function(
                     new CacheEventConverterAsConverter<>(converter))));
            }
         }
```

Reply:
Sure